### PR TITLE
feat: add quotations around filenames with spaces. exa pr#1165

### PR DESCRIPTION
**Pull request by PThorpe92**
https://github.com/eza-community/eza/issues/311

the commit history was a mess on the original PR, and its time to revisit this :D 

image
 
feedback/testers needed

### CODE REVIEW

**Comment by cafkafk** [comment]
Wait, are we doing this before #195, not sure if we came to any conclusions on element.

**Comment by PThorpe92** [comment]
> Wait, are we doing this before #195, not sure if we came to any conclusions on element.

This came about from the referenced issue + digging up a couple of the remaining exa PR's.. I don't remember this one being the one we were waiting on clap for, but that is perfectly fine with me either way as I think it's a good 1.0 addition.

**Comment by gierens** [comment]
Just tested this locally, looks solid ... but I agree we should wait for clap and the properly review this.

**Comment by ariasuni** [comment]
I think quotes should use a different color when they’re not part of the filename but added by exa, so that we can see immediately that they’re not part of the filename

**Comment by gierens** [comment]
> I think quotes should use a different color when they’re not part of the filename but added by exa, so that we can see immediately that they’re not part of the filename

Uh that's ah really nice idea :o :)

**Comment by cafkafk** [comment]
Maybe we shouldn't block this until clap is done, given how big of a task that turned out to be?

File: src/options/help.rs [hunk 1]
```diff
@@ -24,9 +24,11 @@ DISPLAY OPTIONS
   --colo[u]r-scale   highlight levels of file sizes distinctly
   --icons            display icons
   --no-icons         don't display icons (always overrides --icons)
+  --no-quotes        don't quote file names with spaces
   --hyperlink        display entries as hyperlinks
   -w, --width COLS   set screen width in columns
 
+
```
Comment (gierens) [inline src/options/help.rs:31]: why the extra blank line?

File: src/options/flags.rs [hunk 2]
```diff
@@ -15,6 +16,8 @@ pub static TREE:        Arg = Arg { short: Some(b'T'), long: "tree",        take
 pub static CLASSIFY:    Arg = Arg { short: Some(b'F'), long: "classify",    takes_value: TakesValue::Forbidden };
 pub static DEREF_LINKS: Arg = Arg { short: Some(b'X'), long: "dereference", takes_value: TakesValue::Forbidden };
 pub static WIDTH:       Arg = Arg { short: Some(b'w'), long: "width",       takes_value: TakesValue::Necessary(None) };
+pub static NO_QUOTES:Arg = Arg { short: None,          long: "no-quotes",takes_value: TakesValue::Forbidden };
+
```
Comment (gierens) [inline src/options/flags.rs:20]: Don't need the extra blank line here.

File: src/options/flags.rs [hunk 3]
```diff
@@ -80,7 +83,8 @@ pub static ALL_ARGS: Args = Args(&[
     &VERSION, &HELP,
 
     &ONE_LINE, &LONG, &GRID, &ACROSS, &RECURSE, &TREE, &CLASSIFY, &DEREF_LINKS,
-    &COLOR, &COLOUR, &COLOR_SCALE, &COLOUR_SCALE, &WIDTH,
+    &COLOR, &COLOUR, &COLOR_SCALE, &COLOUR_SCALE, &WIDTH, &NO_QUOTES,
+
```
Comment (gierens) [inline src/options/flags.rs:87]: Here too, no extra blank line needed.

File: src/output/file_name.rs [hunk 4]
```diff
@@ -112,6 +116,19 @@ pub enum EmbedHyperlinks{
...
+pub enum QuoteStyle {
+
+    /// Don't ever quote file names.
+    NoQuotes,
+
+    /// Use single quotes for file names that contain spaces and no single quotes
+    /// Use double quotes for file names that contain single quotes.
+    QuoteSpaces,
+}
+
```
Comment (gierens) [inline src/output/file_name.rs:130]: Too many blanks

File: src/options/help.rs (hunk 1 above)
Comment (gierens) [inline src/options/help.rs:31]: too many blanks

File: src/options/file_name.rs [hunk 5]
```diff
@@ -2,16 +2,22 @@ use crate::options::{flags, OptionsError, NumberSource};
...
         let classify = Classify::deduce(matches)?;
         let show_icons = ShowIcons::deduce(matches, vars)?;
-        let embed_hyperlinks = EmbedHyperlinks::deduce(matches)?;
 
-        Ok(Self { classify, show_icons, embed_hyperlinks })
+        let quote_style = QuoteStyle::deduce(matches)?;
+        let embed_hyperlinks = EmbedHyperlinks::deduce(matches)?; 
+
+         Ok(Self { classify, show_icons, quote_style, embed_hyperlinks  })
+
```
Comment (gierens) [inline src/options/file_name.rs:19]: too many blanks

File: src/options/file_name.rs [hunk 6]
```diff
@@ -46,6 +52,18 @@ impl ShowIcons {
     }
 }
 
+
```
Comment (gierens) [inline src/options/file_name.rs:55]: too many blanks

**Review by gierens — CHANGES_REQUESTED** [review]
LGTM :) ... just marked a few formatting nitpicks :D

**Comment by PThorpe92** [comment]
@gierens  those lines were all left over from stripping the other functionality from the PR and fixing merge conflicts, just to give some context. fixing now :D

**Comment by sevz17** [comment]
> I think quotes should use a different color when they’re not part of the filename but added by exa, so that we can see immediately that they’re not part of the filename
> 

That's a good idea, but what happens when you use `--color=never`?

**Comment by gierens** [comment]
> @gierens those lines were all left over from stripping the other functionality from the PR and fixing merge conflicts, just to give some context. fixing now :D

Yeah makes total sense ... I mean without rustfmt such things are bound to happen ... that's why I made the PR for that. :D

**Comment by ariasuni** [comment]
> > I think quotes should use a different color when they’re not part of the filename but added by exa, so that we can see immediately that they’re not part of the filename
> 
> That's a good idea, but what happens when you use `--color=never`?

`ls` align files differently, I think it’s a great idea but not sure how easy it is to do it in our current code base.

```
$ ls -l
total 52
drwxr-xr-x 15 ariasuni ariasuni  4096  6 sept. 16:10  administratif/
drwxr-xr-x  9 ariasuni ariasuni  4096 17 sept. 01:09  documents/
drwxr-xr-x 12 ariasuni ariasuni  4096  6 juil. 13:33  images/
drwxr-xr-x 21 ariasuni ariasuni  4096 28 août  01:33  jeux/
drwxr-xr-x  6 ariasuni ariasuni  4096 22 sept. 18:58  musique/
drwxr-xr-x 11 ariasuni ariasuni  4096 17 juil.  2021  projets/
drwxr-xr-x  5 ariasuni ariasuni  4096 22 sept. 20:14  téléchargements/
-rw-r--r--  1 ariasuni ariasuni     0 22 sept. 23:06 'test file'
drwxr-xr-x  6 ariasuni ariasuni 20480 22 sept. 20:14  vidéos/
```

**Review by cafkafk — APPROVED** [review]
Sandboxed integration tests pass, and this looks good to me when locally testing.


The `--no-quotes` flag also works


Fish and zsh completions also work, and the man pages are also correct. Nice change, LGTM :+1:

**Comment by cafkafk** [comment]
> I think it’s a great idea but not sure how easy it is to do it in our current code base.

Could we do a hacky `\r` prefixed to the first quote? I'm not entirely sure how the rendering code works, or if we wanna actually change the grid code to do this.

**Review by gierens — APPROVED** [review]

**Comment by PThorpe92** [comment]
Had to fix merge conflicts with main.. idk why tests didn't pass. seems to still work fine locally. I'll look at this deeper

**Comment by MartinFillon** [comment]
For the unit tests action your code is not formatted and as for nix the code doesn't follow nix flake check guidelines.

**Review by cafkafk — APPROVED** [review]
Just checked out the changes, still seem good to me.

**Comment by eggbean** [comment]
I know `ls` does this, but what is the practical reason for it? It just looks messier to me.

**Comment by gierens** [comment]
> I know `ls` does this, but what is the practical reason for it? It just looks messier to me.

There are several cases where this can help ... to actually see trailing spaces, to directly copy paste those filenames into further commands without having to add the quotes manually, and of course also for piping eza output into other things ...

If you don't want to see the quotes to can use the `--no-quotes` option.

**Comment by aarondill** [comment]
This pr breaks grid layout when files with spaces are included! Though seemly only when grid and icons are included...
Here is a repro for this bug: https://gist.github.com/aarondill/bf4d306a1819f1c3acb81d5b2a8f7629
Notice the alignment of the first row (the one with the spaces) is messed up when using quotes and icons:

```shell
 > echo Broken with quotes and icons
Broken with quotes and icons
> eza --grid --icons -- /tmp/tmp.Hc5SgTEG8i
 'a file with spaces'   long-file-name-5   long-file-name-10   long-file-name-15   long-file-name-20   long-file-name-25   long-file-name-30   long-file-name-35   long-file-name-40
 long-file-name-1     long-file-name-6   long-file-name-11   long-file-name-16   long-file-name-21   long-file-name-26   long-file-name-31   long-file-name-36
 long-file-name-2     long-file-name-7   long-file-name-12   long-file-name-17   long-file-name-22   long-file-name-27   long-file-name-32   long-file-name-37
 long-file-name-3     long-file-name-8   long-file-name-13   long-file-name-18   long-file-name-23   long-file-name-28   long-file-name-33   long-file-name-38
 long-file-name-4     long-file-name-9   long-file-name-14   long-file-name-19   long-file-name-24   long-file-name-29   long-file-name-34   long-file-name-39
> echo It works without quotes
It works without quotes
> eza --grid --icons --no-quotes -- /tmp/tmp.Hc5SgTEG8i
 a file with spaces   long-file-name-5   long-file-name-10   long-file-name-15   long-file-name-20   long-file-name-25   long-file-name-30   long-file-name-35   long-file-name-40
 long-file-name-1     long-file-name-6   long-file-name-11   long-file-name-16   long-file-name-21   long-file-name-26   long-file-name-31   long-file-name-36
 long-file-name-2     long-file-name-7   long-file-name-12   long-file-name-17   long-file-name-22   long-file-name-27   long-file-name-32   long-file-name-37
 long-file-name-3     long-file-name-8   long-file-name-13   long-file-name-18   long-file-name-23   long-file-name-28   long-file-name-33   long-file-name-38
 long-file-name-4     long-file-name-9   long-file-name-14   long-file-name-19   long-file-name-24   long-file-name-29   long-file-name-34   long-file-name-39
> echo It seems to work without icons tho...
It seems to work without icons tho...
> eza --grid -- /tmp/tmp.Hc5SgTEG8i
'a file with spaces'  long-file-name-5  long-file-name-10  long-file-name-15  long-file-name-20  long-file-name-25  long-file-name-30  long-file-name-35  long-file-name-40
long-file-name-1      long-file-name-6  long-file-name-11  long-file-name-16  long-file-name-21  long-file-name-26  long-file-name-31  long-file-name-36
long-file-name-2      long-file-name-7  long-file-name-12  long-file-name-17  long-file-name-22  long-file-name-27  long-file-name-32  long-file-name-37
long-file-name-3      long-file-name-8  long-file-name-13  long-file-name-18  long-file-name-23  long-file-name-28  long-file-name-33  long-file-name-38
long-file-name-4      long-file-name-9  long-file-name-14  long-file-name-19  long-file-name-24  long-file-name-29  long-file-name-34  long-file-name-39
```

**Comment by aarondill** [comment]
Sorry for the ping. I didn't see #515 or #547.

**Comment by DenebTM** [comment]
oh cool, my PR finally did get merged, without credit but alas :3

> I know `ls` does this, but what is the practical reason for it? It just looks messier to me.

My motivation was: KDE's Konsole relies on the quotes for shell integration stuff, like hovering, CTRL+clicking, or highlighting file names by double click.

**Comment by PThorpe92** [comment]
>oh cool, my PR finally did get merged, without credit but alas :3

Yeah looks like that is my fault, I think the original PR cafk moved over from exa had ping'd you as the author, but that ended up sitting around forever on an old branch so I had to fix so many conflicts and whatnot that I put in a new PR, and only put the link ( `exa PR#1165` ) in the name. 

My apologies, and thank you for the contribution :heart:

**Comment by John-Gee** [comment]
I think the different color for eza quotes from @ariasuni  is a great idea, cause as it is it's hard to see at a glance.

Also is it possible to have the same type of quotes for all? For instance if a folder uses single quotes in its filename eza will add double quote for that folder but will still use single quotes for the rest, it's kind of messy.

Quotes also break alignment I'd say, is it possible to add a spacing to those that don't need quotes, or maybe someone has a better idea for that? Adding quotes to all? Not sure.

**Comment by cafkafk** [comment]
@John-Gee It would probably be a good idea to open an issue or discussion for this, that way your feedback wont be lost!

**Comment by John-Gee** [comment]
@cafkafk very well, done with #728 thank you!

**Pull request merged**